pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,455 @@
+	if err != nil {
+		return nil, nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, generateTag(string(machine.UID)))
+	if err != nil {
+		return nil, nil, err
+	}
+	return device, client, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, _, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.APIKey == "" {
+		return errors.New("apiKey is missing")
+	}
+	if c.InstanceType == "" {
+		return errors.New("instanceType is missing")
+	}
+	if c.ProjectID == "" {
+		return errors.New("projectID is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.APIKey)
+
+	if len(c.Facilities) == 0 || c.Facilities[0] == "" {
+		return fmt.Errorf("must have at least one non-blank facility")
+	}
+
+	// get all valid facilities
+	facilities, _, err := client.Facilities.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list facilities: %v", err)
+	}
+	// ensure our requested facilities are in those facilities
+	if missingFacilities := itemsNotInList(facilityProp(facilities, "Code"), c.Facilities); len(missingFacilities) > 0 {
+		return fmt.Errorf("unknown facilities: %s", strings.Join(missingFacilities, ","))
+	}
+
+	// get all valid plans a.k.a. instance types
+	plans, _, err := client.Plans.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list instance types / plans: %v", err)
+	}
+	// ensure our requested plan is in those plans
+	validPlanNames := planProp(plans, "Name")
+	if missingPlans := itemsNotInList(validPlanNames, []string{c.InstanceType}); len(missingPlans) > 0 {
+		return fmt.Errorf("unknown instance type / plan: %s, acceptable plans: %s", strings.Join(missingPlans, ","), strings.Join(validPlanNames, ","))
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData, userdata string) (instance.Instance, error) {
+	c, _, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	imageName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	serverCreateOpts := &packngo.DeviceCreateRequest{
+		Hostname:     machine.Spec.Name,
```

Comment:
Does the hostname have to be unique per project?

Reply:
No, it does not. Why?

Reply:
And to verify, I just created two servers with the exact same hostname.

Reply:
Perfect